### ISSUE TITLE
`league/players` endpoint optional `week` parameter does not work as expected

### ISSUE THREAD

**Issue by stoopkidddd** [whatadewitt/yahoo-fantasy-sports-api#122]
Only tested this for NFL, so it might be unique to an NFL league since they do the week thing a bit different. When you call the `league.players` endpoint with a `league_key` and `player_key` and additionally a `week` value such as `1`, it ignores the week and just returns the total season points/stats instead.

You can reproduce this in the docs tester. I attached a sample response and screenshot from the tester, but you'd need to use a league that you are in since this is a private league.

![Screenshot 2024-10-26 at 12 48 20 PM](https://github.com/user-attachments/assets/6e3e9b2c-8d35-4842-87b6-f34877d5e32b)

[response.json](https://github.com/user-attachments/files/17531367/response.json)

This does however work on the `player.stats` endpoint, so it seems to be something about the league specificity. I see that `week: "1"` gets sent in the request payload, so I wonder if this is unfortunately just a Yahoo limitation not supporting weekly stats per league 🤔

**Comment by whatadewitt** [comment]
This may be a design of Yahoo! to keep some of their stuff proprietary... but I can dig in more a bit later this week to see why the coverage type is being ignored and where the difference(s) may be. Thanks for pointing it out!

**Comment by whatadewitt** [comment]
I could swear I had something like this working at some time, but I see a number of the issues that seem to indicate exactly what you're saying, it only works on the player endpoint... and my guess is that you're looking for additional stats that are league dependant because I don't believe they will give you every single stat, just a selection of stats that is missing what you're looking for? I can try to prod my contacts at Yahoo! but it's been a while...

**Comment by stoopkidddd** [comment]
It's mainly because I wanted to get a per week breakdown of players' points within the context of a league and didn't want to do the calculation myself. The alternative, which has some benefits over using the `league.players` endpoint, is that I can cache the `player.stats` response for each player in a KV, and then use the league settings to calculate points from the player stats on a per week basis. This means I only need to get the player stats once, instead of once per league.

All in all, not the end of the world, but it would certainly be nice to be able to get a player's calculated points for a given week within the context of a league.

**Comment by whatadewitt** [comment]
100%

I've reached out to someone I know at Yahoo!

In baseball this is especially frustrating because it won't bring back all of the stats, and if you play in a league that uses, as an example, OPS, there's not really a way to get that.

Hopefully I can get some good news from them.

**Comment by alperenekin** [comment]
For basketball, in none of the endpoints I can get the weekly stats. I assume that's an issue on Yahoo's end.
When I try to get **player.stats.** the response says for the stats :   "stats": "Cannot retrieve player stats of type 'week' for game '454'"

I wonder why it is like that.

**Comment by whatadewitt** [comment]
I think it's just because it's proprietary to Yahoo! And fantasy. The stats themselves aren't, but the way your lineup was set is, and I'm guessing they don't want to expose it because it's clearly visible when looking at the matchup pages.

**Comment by alperenekin** [comment]
Seems like that yes. I try to find any other way to get average stats. It doesn't have to be weekly average but I do not see any way to get average stats. Only season total exists as far as I see.

**Comment by whatadewitt** [comment]
So you could use the league settings resource to grab the week information (start / end) and then via the roster.players resource with the "stats" subresource get the stats for each individual day and do some of the calculations yourself. I just tried grabbing last nights stats for my team and I see this, which matches...

```
{
      "player_key": "458.p.10166",
      "player_id": "10166",
      "name": {
        "full": "Willson Contreras",
        "first": "Willson",
        "last": "Contreras",
        "ascii_first": "Willson",
        "ascii_last": "Contreras"
      },
      "url": "https://sports.yahoo.com/mlb/players/10166",
      "editorial_player_key": "mlb.p.10166",
      "editorial_team_key": "mlb.t.24",
      "editorial_team_full_name": "St. Louis Cardinals",
      "editorial_team_abbr": "STL",
      "editorial_team_url": "https://sports.yahoo.com/mlb/teams/st-louis/",
      "is_keeper": {
        "status": false,
        "cost": "",
        "kept": false
      },
      "uniform_number": "40",
      "display_position": "C,1B",
      "headshot": {
        "url": "https://s.yimg.com/iu/api/res/1.2/hNOT688auCzkd2nPBc3Ulw--~C/YXBwaWQ9eXNwb3J0cztjaD0yMzM2O2NyPTE7Y3c9MTc5MDtkeD04NTc7ZHk9MDtmaT11bGNyb3A7aD02MDtxPTEwMDt3PTQ2/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/04232024/10166.png",
        "size": "small"
      },
      "image_url": "https://s.yimg.com/iu/api/res/1.2/hNOT688auCzkd2nPBc3Ulw--~C/YXBwaWQ9eXNwb3J0cztjaD0yMzM2O2NyPTE7Y3c9MTc5MDtkeD04NTc7ZHk9MDtmaT11bGNyb3A7aD02MDtxPTEwMDt3PTQ2/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/04232024/10166.png",
      "is_undroppable": "0",
      "position_type": "B",
      "primary_position": "C",
      "eligible_positions": [
        "C",
        "1B",
        "Util"
      ],
      "eligible_positions_to_add": [],
      "has_player_notes": 1,
      "player_notes_last_timestamp": 1747456233,
      "selected_position": "C",
      "starting_status": 1,
      "batting_order": "3",
      "player_stats": {
        "coverage_type": "date",
        "coverage_value": "2025-05-21",
        "stats": [
          {
            "stat_id": "60",
            "value": "1/3"
          },
          {
            "stat_id": "7",
            "value": "0"
          },
          {
            "stat_id": "12",
            "value": "0"
          },
          {
            "stat_id": "13",
            "value": "1"
          },
          {
            "stat_id": "16",
            "value": "0"
          },
          {
            "stat_id": "3",
            "value": ".333"
          }
        ]
      }
    },
```

Would that help?

**Comment by alperenekin** [comment]
Yes I agree this can be a solution. Only issue that it requires many calls which might be slow to compute the output. Still worth to try!

**Comment by whatadewitt** [comment]
You will get rate limited, but you can run a lot of the calls in parallel, so it's worth a shot!

**Comment by alperenekin** [comment]
Agree. I also see as a second option there are public apis available which at least give you season total data with game played so possible to calculate season average.

**Comment by alperenekin** [comment]
I found out that you can use roster resource.players and actually get weekly total data for given team's each player.

![Image](https://github.com/user-attachments/assets/d582bb8c-3e43-4e3d-9498-3eaa60a2eaa2)

In my case this seems to be helpful!

**Comment by drewffxi** [comment]
I know this is a necro bump, but THANK YOU. I was going crazy because I couldn't find weekly player stats anywhere else. I could get seasonal or 'lastweek'  but could not get it by week. this is SO helpful.

> I found out that you can use roster resource.players and actually get weekly total data for given team's each player.
> 
> ![Image](https://github.com/user-attachments/assets/d582bb8c-3e43-4e3d-9498-3eaa60a2eaa2)
> 
> In my case this seems to be helpful!

**Comment by drewffxi** [comment]
Hi, whatadewitt

Sorry for the possibly stupid question.

I'm writing because I'm hoping you can shed some light on a specific API call that has me completely stumped. I am developing an application in Python, and I'm trying to fetch the weekly stats for all players on a team's roster.

Your documentation site and API tester prove that this is possible. When I use your web tool to make a call to the /roster/players endpoint with a specific week, the JSON response correctly includes the player_stats block with the weekly stats for every player. This is the exact functionality I need.

However, I have been completely unable to replicate this result in Python. I have tried multiple methods, including:

Using the yahoo_fantasy_api Python library.

Making a direct, raw HTTP GET request to the official Yahoo API endpoint (https://fantasysports.yahooapis.com/fantasy/v2/team/{team_key}/roster;week={week_num}/players;out=stats).

In every attempt from my Python script, the API returns a valid response, but the crucial player_stats block is always missing from the player data. It seems my direct requests are being treated differently by Yahoo's servers than the requests originating from your application.

My question is: Could you possibly share any insights into how your application's backend successfully makes this call?

Thank you again for your time and for the fantastic work you've done.

**Comment by whatadewitt** [comment]
So the code for that function is here: https://github.com/whatadewitt/yahoo-fantasy-sports-api/blob/master/resources/rosterResource.mjs#L42-L75

Trying to parse through it I think you want something like

`https://fantasysports.yahooapis.com/fantasy/v2/team/${teamKey}/roster/players/stats;type=week&week=${week_num}`

Now I do notice that on my docs site that is currently returning an error. I have been working on rebuilding this thing in TypeScript when time allows (note: I don't have a ton of time) but I know that all of my endpoints were working, but I am thinking that I may have removed the `roster.players` call from the version because it was no longer working... I will do my best to check on it later today if I am on the machine where I have that code and let you know...

**Comment by whatadewitt** [comment]
@drewffxi just confirmed on that new branch (sorry for the delay... travel for work this week)

https://fantasysports.yahooapis.com/fantasy/v2/team/mlb.l.3555.t.1/roster;week=12/players;out=stats

that is working for me right now... HOWEVER if I do not provide a week it is showing empty stats... going to try season and can report back.

**Comment by whatadewitt** [comment]
It could be a sport by sport thing

https://fantasysports.yahooapis.com/fantasy/v2/team/nfl.l.226418.t.1/roster/players;week=7;out=stats shows me what I want -- the weekly points for the players on my fantasy football team

https://fantasysports.yahooapis.com/fantasy/v2/team/mlb.l.3555.t.1/roster/players;week=13;out=stats -- baseball forces to "date", I cannot get the weekly stats for my player. So you might just have to play with things and make multiple requests in some cases.

**Comment by drewffxi** [comment]
Thanks for the replies. Yeah I pretty much confirmed it can't do weekly stats. Found a workaround by just summing the stats for each individual day for the week. At least you can batch the requests 

Thanks again for reply

**Comment by whatadewitt** [comment]
I used to get replies from some of the folks at Yahoo!, not so much anymore. I would love to have more access to help things like this... maybe one day! Best of luck! Please share your work when you're done, if you can!

**Comment by drewffxi** [comment]
Hey, thank you again for replying.
I honestly don't have any experience with coding or APIs but I wanted to make a Newsletter Generator for my fantasy hockey league, so I used AI to help me make this webapp thing:
https://FantasyNewsletterPRO.com

Am stumbling through it, has been a fun little project to work on and I've learned a lot about stuff I never thought I would.

And I really appreciate you taking the time to get back to me. Your website was really insightful. Thanks again

**Comment by whatadewitt** [comment]
Oh that's fantastic. I do the same for my fantasy baseball league, and use Notebook LLM to make a podcast for the group too!

Best of luck!

**Comment by drewffxi** [comment]
Woah, what do you mean you make a podcast?! That sounds so cool.

Edit: nevermind! Figured it out. Thanks for the neat tip!
